java/ql/src/experimental/Security/CWE/CWE-749/UnsafeAndroidAccess.ql

```diff
@@ -0,0 +1,135 @@
+import semmle.code.java.frameworks.android.WebView
+import semmle.code.java.dataflow.FlowSources
+
+/**
+ * Methods allowing any-local-file and universal access in the WebSettings class
+ */
+class CrossOriginAccessMethod extends Method {
+  CrossOriginAccessMethod() {
+    this.getDeclaringType() instanceof TypeWebSettings and
+    (
+      this.hasName("setAllowUniversalAccessFromFileURLs") or
+      this.hasName("setAllowFileAccessFromFileURLs")
+    )
+  }
+}
+
+/**
+ * `setJavaScriptEnabled` method for the webview
+ */
+class AllowJavaScriptMethod extends Method {
+  AllowJavaScriptMethod() {
+    this.getDeclaringType() instanceof TypeWebSettings and
+    this.hasName("setJavaScriptEnabled")
+  }
+}
+
+/**
+ * Holds if `ma` is a method invocation against `va` and `va.setJavaScriptEnabled(true)` occurs elsewhere in the program
+ */
+predicate isJSEnabled(Variable v) {
+  exists(VarAccess va, MethodAccess jsa |
+    v.getAnAccess() = va and
+    jsa.getQualifier() = v.getAnAccess() and
+    jsa.getMethod() instanceof AllowJavaScriptMethod and
+    jsa.getArgument(0).(BooleanLiteral).getBooleanValue() = true
+  )
+}
+
+/**
+ * Fetch URL method call on the `android.webkit.WebView` object
+ */
+class FetchResourceMethodAccess extends MethodAccess {
+  FetchResourceMethodAccess() {
+    this.getMethod().getDeclaringType() instanceof TypeWebView and
+    (
+      this.getMethod().hasName("loadUrl") or
+      this.getMethod().hasName("postUrl")
+    )
+  }
+}
+
+/**
+ * Method access to external inputs of `android.content.Intent` object
+ */
+class IntentGetExtraMethodAccess extends MethodAccess {
+  IntentGetExtraMethodAccess() {
+    this.getMethod().getName().regexpMatch("get\\w+Extra") and
+    this.getMethod().getDeclaringType() instanceof TypeIntent
+    or
+    this.getMethod().getName().regexpMatch("get\\w+") and
+    this.getQualifier().(MethodAccess).getMethod().hasName("getExtras") and
+    this.getQualifier().(MethodAccess).getMethod().getDeclaringType() instanceof TypeIntent
+  }
+}
+
+/**
+ * Source of fetching urls
+ */
+class UntrustedResourceSource extends RemoteFlowSource {
+  UntrustedResourceSource() {
+    exists(IntentGetExtraMethodAccess ma |
+      this.asExpr().(VarAccess).getVariable().getAnAssignedValue() = ma
+    )
+  }
+
+  override string getSourceType() {
+    result = "user input vulnerable to XSS and sensitive resource disclosure attacks" and
+    exists(MethodAccess ma |
+      ma.getMethod() instanceof CrossOriginAccessMethod and //High precision match of unsafe resource fetching
+      ma.getArgument(0).(BooleanLiteral).getBooleanValue() = true
+    )
+    or
+    result = "user input potentially vulnerable to XSS and sensitive resource disclosure attacks" and
+    not exists(MethodAccess ma |
+      ma.getMethod() instanceof CrossOriginAccessMethod and //High precision match of unsafe resource fetching
+      ma.getArgument(0).(BooleanLiteral).getBooleanValue() = true
+    )
+  }
+}
```

Comment:
```suggestion
  predicate crossOriginAccessEnabled() {
    exists(MethodAccess ma |
      ma.getMethod() instanceof CrossOriginAccessMethod and // High precision match of unsafe resource fetching
      ma.getArgument(0).(BooleanLiteral).getBooleanValue() = true
    )
  }

  override string getSourceType() {
    if crossOriginAccessEnabled() 
    then result = "user input vulnerable to XSS and sensitive resource disclosure attacks"
    else result = "user input potentially vulnerable to XSS and sensitive resource disclosure attacks"
  }
}
```

Reply:
Although, am I reading this right that it will regard cross-origin access as enabled if there is *any* `xyz.setAllowFileAccessFromFileURLs(true)` in the whole program? That doesn't seem very precise.

Reply:
Thanks @smowton , this is an excellent point. I've moved the code to the sink part and added the check that the qualifier <code>xyz</code> of the type <code>WebSettings</code> is the same variable accessed by the <code>WebView</code> object that invokes unsafe URL fetching.

Now it is precise and is exactly what the query needs.